servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/HttpRequestAutoDrainingServiceFilter.java

```diff
@@ -0,0 +1,161 @@
+
+    private HttpRequestAutoDrainingServiceFilter() {
+        // singleton
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    @Override
+    public StreamingHttpServiceFilter create(StreamingHttpService service) {
+        return new DrainingStreamingHttpServiceFilter(service);
+    }
+
+    private static final class DrainingStreamingHttpServiceFilter extends StreamingHttpServiceFilter {
+
+        DrainingStreamingHttpServiceFilter(StreamingHttpService service) {
+            super(service);
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> handle(HttpServiceContext ctx, StreamingHttpRequest request,
+                                                    StreamingHttpResponseFactory responseFactory) {
+            final DrainTerminalSignalConsumer terminalSignalConsumer = new DrainTerminalSignalConsumer(request);
+            request.transformMessageBody(body ->
+                    body.beforeSubscriber(() -> {
+                        terminalSignalConsumer.requestSubscribed();
+                        return NoopSubscriber.INSTANCE;
+                    }));
+            return delegate().handle(ctx, request, responseFactory)
+                    .liftSync(new AfterFinallyHttpOperator(terminalSignalConsumer, true));
+        }
+    }
+
+    private static final class DrainTerminalSignalConsumer implements TerminalSignalConsumer {
+
+        private static final AtomicIntegerFieldUpdater<DrainTerminalSignalConsumer> UPDATER =
+                AtomicIntegerFieldUpdater.newUpdater(DrainTerminalSignalConsumer.class, "state");
+
+        private static final int PENDING = 0;
+        private static final int COMPLETE = 1;
+
+        private final StreamingHttpRequest request;
+        @SuppressWarnings("unused")
+        private volatile int state = PENDING;
+
+        DrainTerminalSignalConsumer(StreamingHttpRequest request) {
+            this.request = request;
+        }
+
+        void requestSubscribed() {
+            once();
+        }
+
+        @Override
+        public void onComplete() {
+            if (once()) {
+                request.payloadBody().ignoreElements().shareContextOnSubscribe().subscribe();
+            }
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            cancel(); // same behavior as cancel()
+        }
+
+        @Override
+        public void cancel() {
+            if (once()) {
+                toSource(request.messageBody().shareContextOnSubscribe())
+                        .subscribe(CancelImmediatelySubscriber.INSTANCE);
+            }
+        }
+
+        private boolean once() {
+            return UPDATER.compareAndSet(this, PENDING, COMPLETE);
+        }
+    }
+
+    private static final class NoopSubscriber implements PublisherSource.Subscriber<Object> {
```

Comment:
This is copied over from elsewhere: do we have any centralized versions of this?